bpf/bpf_lxc.c

```diff
@@ -1941,6 +1941,119 @@ int tail_ipv4_policy(struct __ctx_buff *ctx)
 	return ret;
 }
 
+static __always_inline bool
+ipv4_to_endpoint_is_hairpin_flow(struct __ctx_buff *ctx, struct iphdr *ip4, __u32 __maybe_unused src_sec_identity)
+{
+	struct ipv4_ct_tuple tuple = {};
+	struct lb4_backend *backend __maybe_unused;
+	struct ct_entry *entry;
+	struct ct_map *map;
+	int err, l4_off;
+	__be32 pod_ip, service_ip;
+	__be16 client_port, backend_port, service_port;
+
+	/* For reasons unknown, hairpinned packets don't have their IDs resolved and
+	 * are always seen as originating from world.
+	 */
+	if (src_sec_identity != WORLD_ID)
+		return false;
```

Comment:
This reads as if the result is unintended; an integration test will be necessary to prevent a (possible) fix for ID resolution from rendering this function moot.

Reply:
Agreed, in fact, after merging https://github.com/cilium/cilium/pull/27602 and rebasing, we discovered that packets in the reply direction suddenly _did_ receive proper secID and succeeded the policy check. This means a skip was no longer necessary, so no point in the hairpin check to begin with, but you're right in that it's not the most robust solution. I've removed the check to let the CT lookups do the work.